lib/src/analysis_options/io_file_system.dart

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+
+import 'file_system.dart';
+
+/// An implementation of [FileSystem] using `dart:io`.
+class IOFileSystem implements FileSystem {
+  Future<IOFileSystemPath> makePath(String path) async =>
+      IOFileSystemPath._(path);
+
+  @override
+  Future<bool> fileExists(FileSystemPath path) => File(path.ioPath).exists();
+
+  @override
+  Future<FileSystemPath> join(FileSystemPath from, String to) =>
+      makePath(p.join(from.ioPath, to));
+
+  @override
+  Future<FileSystemPath?> parentDirectory(FileSystemPath path) async {
+    // Make [path] absolute (if not already) so that we can walk outside of the
+    // literal path string passed.
+    var result = p.dirname(p.absolute(path.ioPath));
+
+    // If the parent directory is the same as [path], we must be at the root.
+    if (result == path.ioPath) return null;
+
+    return makePath(result);
+  }
+
+  @override
+  Future<String> readFile(FileSystemPath path) =>
+      File(path.ioPath).readAsString();
+}
+
+/// An abstraction over a file path string, used by [IOFileSystem].
+///
+/// To create an instance of this, use [IOFileSystem.makePath()].
+class IOFileSystemPath implements FileSystemPath {
+  /// The underlying physical file system path.
+  final String path;
+
+  IOFileSystemPath._(this.path);
+}
+
+extension on FileSystemPath {
+  String get ioPath => (this as IOFileSystemPath).path;
```

Comment:
Would it make sense to add a generic to `FileSystem` so that we don't need to cast `FileSystemPath` instances?

Or can we use `covariant` to make the expectation visible in the signature?

Reply:
I did consider that. I believe this is a somewhat famous design problem in programming languages. The problem is if you do that, all code that works with `FileSystem` has to itself become generic. You end up having to propagate the type parameter through entire regions of the program.

That's not what you want because most other code doesn't care about the concrete `FileSystemPath` type. The feature we would actually need to model this without using casts is abstract types, like Rust, ML and some other languages have. Something like:

```dart
abstract class FileSystem {
  abstract type Path;

  ...
}

class IOFileSystem implements FileSystem {
  type Path = IOFileSystemPath;

  ...
}

class IOFileSystemPath implements FileSystemPath { ... }
```

But Dart doesn't have that. OOP works really well for modeling a *single* hierarchy of types with encapsulation, but doesn't handle families of related types like this. "Family polymorphism" is the name for another feature that I think tackles these kinds of problems.

In the absence of that, I just went with downcasting. I like your suggestion to mark the method parameters with `covariant`, though. I'll do that.